### PR TITLE
Implemented an improved replyTo for IsCompleteRequests

### PR DESCRIPTION
**Pull request by kfiz**
To determine wether a line of code is complete or not, we parse the codestring in the IsCompleteRequest and assume the line is incomplete if an error occurs.
This works for the qtconsole but not for the console. 
The latter doesn't send is_complete_requests at present, see
https://github.com/jupyter/jupyter_console/issues/11

Ping @gibiansky @sumitsahrawat

### CODE REVIEW

File: src/IHaskell/Eval/Evaluate.hs [hunk 1]
```diff
@@ -264,10 +265,17 @@ data EvalOut =
...
-                  else str
+                  else s
   where
     str = strip x
-    l = lines str
+    -- check wether the string still has a newline from continuation
+    hasNewline = elem '\n' str
+    s =
+      case hasNewline of
+        True -> foldr (++) [] $ split "\n" $ str
```
Comment (sumitsahrawat) [inline src/IHaskell/Eval/Evaluate.hs]: Looks like this can be replaced by `T.filter (/= '\n') str`.

**Comment by sumitsahrawat** [comment]
Everything looks great (except for that small change). Thanks for the patch :+1:

Would you be willing to look into the build faliures? There would also be some formatting issues, which can be fixed by using [gibiansky/hindent](https://github.com/gibiansky/hindent) if you want.

It seems like some package has failed builds, if so you can add version constraints to avoid that specific version.

File: src/IHaskell/Eval/Evaluate.hs (hunk 1 above)
Comment (kfiz) [inline src/IHaskell/Eval/Evaluate.hs]: Agreed.

**Comment by kfiz** [comment]
Thx. I'll look into the failed builds. Regarding formatting, which files do I' have to reformat? I actually thought I had formatted the files correctly, as I had already used hindent...

File: src/IHaskell/Eval/Parser.hs [hunk 2]
```diff
@@ -110,7 +110,7 @@ parseString codeString = do
           block <- parseChunk chunk line
           activateExtensions $ unloc block
           processChunks (block : accum) remaining
-
+
```
Comment (sumitsahrawat) [inline src/IHaskell/Eval/Parser.hs]: This is a definite formatting issue for `hindent`.

**Comment by sumitsahrawat** [comment]
Formatting can be verified using the `verify_formatting.py` script in the root directory. Looking at the code, I expect there to be some formatting issues atleast.

**Comment by kfiz** [comment]
Ok guys, I need some support here. 
There has been an upstream update of the `nats`package, the current version is `1.1`.
The problem is that the `nats 1.1.` package declares a class  instance, that leads to a duplicate instance error in the `ghc 7.6` and `ghc 7.8` builds because the package `th-orphans 12.2.2` declares the same class instance. Unfortunately the `nats` package is part of the `base` package. I tried to get cabal to use the `nats 1` package, which does not declare this instance, but this only worked for the `ghc 7.6` build. The changes I made to the `*.cabals` have no effect to the `ghc 7.8` builds.

What is the best way to ensure that the `nats 1` package is pulled for the `ghc 7.6` and `ghc 7.8` builds? Does anybody else see this?

**Comment by sumitsahrawat** [comment]
I fixed it on the master branch. You can rebase your work on that. As you're working on your master branch, a simple `git pull <branch-representing-original-repo> --rebase` should work.

**Comment by sumitsahrawat** [comment]
I think the last four commits can be removed.

**Comment by kfiz** [comment]
I followed your suggestion an have split code input from parsing. So at present multiline input is possible. Fixing the let-statement problem is a bit harder without ghci support, since I yet don' t know what exactly causes the parse errors, that occur when trying to enter properly indented correct Haskell code. 
Obviously all of the parsers here:

https://github.com/gibiansky/IHaskell/blob/master/src/IHaskell/Eval/Parser.hs#L193-L206

fail. Especially parserStatement fails.
In ghci such multiline let-statements are only possible when setting `:set +m`.

**Comment by sumitsahrawat** [comment]
Great, I think there should be a way to apply the `:set +m` effect to the haskell interpreter inside IHaskell. I'll take a look into it later.

**Comment by kfiz** [comment]
Great. To avoid any misunderstandings. Statements like:

    let x = 42
        b = 1
    x + b

are already possible. But

    let x = 42
        b = 1
    in x + b

throws an error. This works in ghci with `:set +m`.
Your example 

    let x = 
      x in x
does not work, even with `:set +m`.
But 

    let x = x
    in x
works with the option set. I think the former might actually violate Haskell indentation rules. But I'm not sure.

**Comment by sumitsahrawat** [comment]
This fell off the stack for me. I completely forgot about it, sorry :smile:

@kfiz Does it work well for you? If so, would you be willing to squash the changes into 2-3 commits so that redundant commits go out of the way? e.g one for the multiline fix, and one for the parsing check?

**Comment by kfiz** [comment]
No problem. What is the cleanest way to go about removing the unnecessary commits and issuing one pull request that has all the changes I made? Should I close the request and issue a new one?

**Comment by sumitsahrawat** [comment]
Don't close this one and open a new PR. This would be a great chance to learn about squashing. Take a look at this for more: http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

I've pulled your changes to my fork, so you can experiment freely :smile: :+1:

**Comment by kfiz** [comment]
Thx for the link. I think you can merge now. I have combined all relevant commits into one.

**Comment by sumitsahrawat** [comment]
Your fork is still in a pretty wild state. The commits tab above shows 17 commits, most of which cancel each other out.

The best way out from this situation would be to make these changes again in a clean branch and force push it. It would have been much easier if you were not working on the master branch.

I found a really simple way of doing this while writing this mini-tutorial, so all's good :smile:.

---

* Add a remote to the original repository.

```bash
git remote add upstream https://github.com/gibiansky/IHaskell
```

* Then we take a look at the current situation:
```
$ git log --oneline --all --graph --decorate
*   a56e7a1 (HEAD -> master, origin/master, origin/HEAD) Merge branch 'master' of https://github.com/kfiz/IHaskell
|\  
| *   bf9ceab Merge branch 'master' of https://github.com/kfiz/IHaskell
| |\  
| | * e14acdf improved replyTo for isCompleteRequests. multiline statements are now possible in the qtconsole. Code has to be entered with a trailing empty line
| | * 1fe6171 Adjusted the comment of the replyTo function for isCompleteRequests. Corrected typos
| | * fc35aff Adjusted the comment of the replyTo function for isCompleteRequests
| | * c2a2b36 Adjusted the comment of the replyTo function for isCompleteRequests
| | * c1255f2 Formatting error fix
| | * 88d0a6d Patch that allows multiline input in the qtconsole.
| | * d820aef lifted the nats restriction, as it is not necessary anymore since the dependencies of the haskell-src-meta package have changed
| * | 6aa5c86 improved replyTo for isCompleteRequests. multiline statements are now possible in the qtconsole. Code has to be entered with a trailing empty line
| * | a5fd820 improved replyTo for isCompleteRequests. multiline statements are now possible in the qtconsole. Code has to be entered with a trailing empty line
| |/  
* | 3839b21 improved replyTo for isCompleteRequests
|/  
* 7b30df9 Downloaded and used original Hspec.hs to verify it is correct
* 9f45159 reversed the changed Hspec import
* 47a7892 Hspec Fix
* 906d8f5 Formatting Fix
* 046a4c8 identical commit to initial commit
| *   16d5ab0 (upstream/master) Merge pull request #609 from bollu/fix-magic-macos-install
| |\  
|/ /  
| * eadd920 use brew --prefix to figure out library and include paths
| * 03e5494 fix macOS cabal magic installation
|/  
* ece7e45 Prevent re-use of ipython-package from hackage
```

* The `HEAD` of the `master` branch is at `origin/master`. We move this to `upstream/master`.

```bash
git reset upstream/master
# Now the HEAD of the fork's master branch is aligned with the HEAD of the upstream
# master branch. This doesn't destroy the changes you made, and you can commit here.
```

* Now we create a new branch (**skip this for now**, as the PR's already open)

```bash
git checkout -b qtconsole-multiline
```

* Check the diff (`git diff`) and commit.

**Comment by kfiz** [comment]
Thx for the tutorial, I followed it and force pushed my changes in a single commit to my fork. Hope everything is ok now.

**Comment by sumitsahrawat** [comment]
Everything is stellar :+1: Thanks for taking the time.

The build issues are due to updated libraries (#615), so I'll go ahead and merge this.

**Pull request merged**